src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,301 @@
+   * should return [[restaurantOne, restaurantTwo]]. Scope of method is public for testing purposes.
+   *
+   * @param nonStreetAddressWaypointsAsPlaceTypes A list of place types to call search for. (e.g.
+   *     restaurant, supermarket, police station)
+   * @param streetAddressesAsCoordinates A list of coordinates to look for the place types around.
+   * @return A list of lists of place IDs where each list represents the search nearby result for
+   *     every known coordinate.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   */
+  public List<List<String>> searchNearbyEveryKnownLocationForClosestPlaceTypeMatch(
+      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
+      List<LatLng> streetAddressesAsCoordinates)
+      throws PlacesException {
+    List<List<String>> allSearchNearbyResults = new ArrayList<>();
+    for (PlaceType nonStreetAddressWaypoint : nonStreetAddressWaypointsAsPlaceTypes) {
+      List<String> searchNearbyResults = new ArrayList<>();
+      for (LatLng coordinate : streetAddressesAsCoordinates) {
+        PlaceType placeType = nonStreetAddressWaypoint;
+        RankBy rankBy = RankBy.DISTANCE;
+        PlacesClient placesClient = placesClientFactory.getPlacesClient(apiKey);
+        String nearestMatch = placesClient.searchNearby(coordinate, placeType, rankBy);
+        if (nearestMatch != null) {
+          searchNearbyResults.add("place_id:" + nearestMatch);
+        }
+      }
+      allSearchNearbyResults.add(searchNearbyResults);
+    }
+    return allSearchNearbyResults;
+  }
+
+  /**
+   * Chooses the combination of waypoints that results in the shortest travel time possible. Scope
+   * of method is public for testing purposes.
+   *
+   * @param origin The starting point of travel.
+   * @param destination The ending point of travel.
+   * @param allWaypointCombinations A list of waypoint combinations to select between for the
+   *     shortest travel time possible.
+   * @return The most optimal combination of waypoint with the shortest travel time.
+   * @throws DirectionsException An exception thrown when an error occurs with the Directions API.
+   */
+  public List<String> chooseWaypointCombinationWithShortestTravelTime(
+      String origin, String destination, List<List<String>> allWaypointCombinations)
+      throws DirectionsException {
+    long minTravelTime = 0;
+    List<String> mostOptimalWaypointCombination = new ArrayList<String>();
+    for (List<String> waypointCombination : allWaypointCombinations) {
+      DirectionsClient directionsClient = directionsClientFactory.getDirectionsClient(apiKey);
+      DirectionsResult directionsResult =
+          directionsClient.getDirections(origin, destination, waypointCombination);
+      long travelTime = DirectionsClient.getTotalTravelTime(directionsResult);
+      if (minTravelTime == 0 || travelTime < minTravelTime) {
```

Comment:
Use `Long.MAX_VALUE` for the initial value of `minTravelTime`, and then you can simplify this check to be `Math.min(minTravelTime, travelTime)`

Reply:
Hmm not sure if I am understanding correctly, I kinda need to reassign `mostOptimalWaypointCombination` so I don't think that would help.

Reply:
It would remove the need to check if `mintravelTime == 0`. In competitive programming, we always do that. As a matter of fact, there's a magic number we use to represent infinity "0x3f3f3f3f". It is "magic" because it is large, doubling it still fits into a 32 bit integer, and because the bytes of it are the same, you can fill an array with that value using the C function `memset`. Though, I'm not sure if that's considered a "good" practice :( Maybe using `Optional` would be preferred.

Reply:
@gbuenoandrade Sounds like a cool number :) +1 on using optional if `Long.MAX_VALUE` is not recommended

Reply:
Optional is a great idea, I think it gives more information. Thanks!

Reply:
minTravelTime.isPresent()

Reply:
`RESOLVED`